lambdas/nino-check/src/main.ts

```
@@ -0,0 +1,116 @@
+  const { logger } = helpers.logHelper;
+
+  const { nino, sessionId, deviceInformationHeader } = processEvent(event);
+
+  const functionConfig = getConfig({ deviceInformationHeader });
+  const { tableNames } = functionConfig;
+
+  logger.info(`Getting person & session data...`);
+
+  let session: SessionItem;
+  let personIdentity: PersonIdentityItem;
+  let isFinalAttempt: boolean;
+
+  try {
+    ({ session, personIdentity, isFinalAttempt } = await getSessionInfo(tableNames, helpers, sessionId));
+  } catch (error) {
+    if (error instanceof RecordExpiredError) {
+      helpers.metricsHelper.captureMetric(`InvalidSessionErrorMetric`);
+      throw new CriError(
+        400,
+        `Expired record(s) in ${error.tableName} (expir[y/ies]: ${error.expiryDates.join(", ")})`
+      );
+    } else if (error instanceof RecordNotFoundError) {
+      throw new CriError(500, `No ${error.tableName} record found with sessionId '${error.sessionId}'`);
+    } else if (error instanceof TooManyAttemptsError) {
+      helpers.metricsHelper.captureMetric(`AttemptsExceededMetric`);
+      logger.info(`Caught TooManyAttemptsError. Returning requestRetry: false.`);
+      return { statusCode: 200, body: JSON.stringify({ requestRetry: false }) };
+    }
+
+    logger.info(`Caught unexpected session info function error: ${String(error)}`);
+
+    throw new CriError(500, "Unexpected error getting session information");
+  }
+
+  const { clientId, clientSessionId: govJourneyId } = session;
+
+  helpers.logHelper.logEntry(govJourneyId);
+
+  logger.info(
+    `Retrieved session and person identity. isFinalAttempt=${isFinalAttempt}. Validating provided NINo information...`
+  );
+
+  let ninoMatch: boolean;
+
+  try {
+    ({ ninoMatch } = await validateNino(clientId, functionConfig, helpers, personIdentity, session, nino));
+  } catch (error) {
+    if (error instanceof PdvApiError) {
```

Do we actually need to throw these custom errors, or can we just throw CRIErrors with the message?
It just looks like we throw one custom error, do nothing with it but catch it and throw a different custom exception?

---

Right ok, I'll always lean towards keeping things simple. I'm not convinced we need such separation within our own codebase, but I'm content for it to be kept like this if you see value in it.

---

Throw CriErrors